Часть третья: Анализ датасет c Kaggle.com#
В этой части домашней работы мы еще больше углубимся в библиотеку pandas и рассмотрим более продвинутые функции.

Для этого упражнения мы используем данные по Топ-50 рейтингу самых продаваемых книг на Amazon за 11 лет (с 2009 по 2019). Датасет находится в открытом доступе на Kaggle.com. Загрузите csv файл по ссылке и переместите его в ту же директорию, где находится ваш рабочий ноутбук (для удобства). После этого переходите к заданию

Для выполнения этой части домашнего задания, нужно будет не только написать код, но и ответить на сопутствующие вопросы. Там, где вы увидите выделенную жирным шрифтом надпись ответ:, нужно будет вставить вопрос в файл и ответ на него.

Например:

Какая библиотека используется для работы с датафреймами в python? Ответ: pandas

Необходимо выполнить:

Прочитайте csv файл (используйте функцию read_csv)
Выведите первый пять строчек (используется функцию head)
Выведите размеры датасета (используйте атрибут shape)
Ответ: О скольки книгах хранит данные датасет?

In [309]:
import os
import pandas as pd
link = "https://www.kaggle.com/sootersaalu/amazon-top-50-bestselling-books-2009-2019"
df=pd.read_csv("bestsellers with categories.csv", encoding='utf8', sep=',')
df.head(5)

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

Выведите размеры датасета (используйте атрибут shape)

In [310]:
df.shape

(550, 7)

Ответ: О скольки книгах хранит данные датасет? **Ответ: 550**

Для каждой из книг доступно 7 переменных(колонок). Давайте рассмотрим их детальней:

Name - название книги

Author - автор

User Rating - рейтинг (по 5-бальной шкале)

Reviews - количество отзывов

Price - цена (в долларах по состоянию на 2020 год)

Year - год, когда книга попала в рейтинг Топ-50

Genre - жанр

Для упрощения последующей работы, давайте немного подправим названия переменных. Как видите, здесь все названия начинаются с большой буквы, а одно - даже содержит пробел. Это крайне нежелательно и может быть достаточно неудобным. Давайте сменим регистр на маленький, а пробел заменим на нижнее подчеркивания (snake_style). А заодно и выучим полезный атрибут датафрейма: columns (можно просто присвоить список новых имен этому атрибуту)

df.columns = ['name', 'author', 'user_rating', 'reviews', 'price', 'year', 'genre']

In [311]:
df.columns = ['name', 'author', 'user_rating', 'reviews', 'price', 'year', 'genre']
df.head(3)

name              author  user_rating  \
0            10-Day Green Smoothie Cleanse            JJ Smith          4.7   
1                        11/22/63: A Novel        Stephen King          4.6   
2  12 Rules for Life: An Antidote to Chaos  Jordan B. Peterson          4.7   

   reviews  price  year        genre  
0    17350      8  2016  Non Fiction  
1     2052     22  2011      Fiction  
2    18979     15  2018  Non Fiction

**Первичное исследование данных**

Проверьте у всех ли строк хватает данных: выведите количество пропусков (na) в каждом из столбцов (используйте функции isna и sum)

In [312]:
for i in df.columns:
    print(f"В колонке {i} доля пропуска составляет {df[i].isnull().sum()}")

В колонке name доля пропуска составляет 0
В колонке author доля пропуска составляет 0
В колонке user_rating доля пропуска составляет 0
В колонке reviews доля пропуска составляет 0
В колонке price доля пропуска составляет 0
В колонке year доля пропуска составляет 0
В колонке genre доля пропуска составляет 0


Ответ: Есть ли в каких либо переменных пропуски? **Ответ: нет**

Проверьте, какие есть уникальные значения в колонке genre (используйте функцию unique)
Ответ: Какие есть уникальные жанры? **Ответ: Non Fiction и Fiction**

In [313]:
df['genre'].unique

<bound method Series.unique of 0      Non Fiction
1          Fiction
2      Non Fiction
3          Fiction
4      Non Fiction
          ...     
545        Fiction
546    Non Fiction
547    Non Fiction
548    Non Fiction
549    Non Fiction
Name: genre, Length: 550, dtype: object>

Определите, какая цена у нас максимальная, минимальная, средняя, медианная (используйте функции max, min, mean, median)
Максимальная цена? **Ответ: 105**
Минимальная цена? **Ответ: 0**
Средняя цена? **Ответ: 13.1**
Медианная цена? **Ответ: 11**

In [314]:
print(df['price'].max(),
df['price'].min(),
df['price'].mean(),
df['price'].median())


105 0 13.1 11.0


In [315]:
print(f'1............. {df["user_rating"].max()}')
print(f'2............. { df[df["user_rating"] == df["user_rating"].max()].shape}')
print(f'3............. { df[df["reviews"] == df["reviews"].max()]}')
df_2015=df[df['year'] == 2015]

print(f'4............. {df_2015[df_2015["price"]==df_2015["price"].max()]}')
print(f'5............. {df[(df["genre"] == "Fiction") & (df["year"] == 2010)].shape}')
all_data = df[(df["user_rating"] == 4.9) & (df["year"].isin([2010, 2011]))]
print(f'6............. {all_data.shape}')



1............. 4.9
2............. (52, 7)
3.............                         name       author  user_rating  reviews  price  year  \
534  Where the Crawdads Sing  Delia Owens          4.8    87841     15  2019   

       genre  
534  Fiction  
4.............                                                   name  \
277  Publication Manual of the American Psychologic...   

                                 author  user_rating  reviews  price  year  \
277  American Psychological Association          4.5     8580     46  2015   

           genre  
277  Non Fiction  
5............. (20, 7)
6............. (1, 7)


**Поиск и сортировка данных#**

1. Какой рейтинг в датасете наивысший? Ответ: **4.9**
2. Сколько книг имеют такой рейтинг? Ответ: **52**
3. У какой книги больше всего отзывов? Ответ: **Where the Crawdads Sing**
4. Из тех книг, что попали в Топ-50 в 2015 году, какая книга самая дорогая (можно использовать промежуточный датафрейм)? ***Ответ:American Psychological Association ***
5. Сколько книг жанра Fiction попали в Топ-50 в 2010 году (используйте &)?**Ответ: 20**
6. Сколько книг c рейтингом 4.9 попало в рейтинг в 2010 и 2011 годах (используйте |или функцию isin)? **Ответ: 1**
 
И напоследок, давайте отсортируем по возрастанию цены все книги, которые попали в рейтинг в 2015 году и стоят дешевле 8 долларов (используйте функцию sort_values).
Какая книга последняя в отсортированном списке? **Ответ: Old School Diary of a Wimpy Kid #10**


In [316]:
df[(df["price"] < 8) & (df["year"] == 2015)].sort_values("price")

name              author  \
507                              To Kill a Mockingbird          Harper Lee   
210                      Laugh-Out-Loud Jokes for Kids         Rob Elliott   
206                         Knock-Knock Jokes for Kids         Rob Elliott   
111                                    First 100 Words        Roger Priddy   
123                               Giraffes Can't Dance       Giles Andreae   
54   Creative Haven Creative Cats Coloring Book (Ad...     Marjorie Sarnat   
63                      Dear Zoo: A Lift-the-Flap Book        Rod Campbell   
89   Dover Creative Haven Art Nouveau Animal Design...         Marty Noble   
478                        The Very Hungry Caterpillar          Eric Carle   
201  Killing Reagan: The Violent Assault That Chang...       Bill O'Reilly   
28                        Baby Touch and Feel: Animals                  DK   
224                                   Love You Forever       Robert Munsch   
55   Creative Haven Owls Coloring Book (Adult Color...     Marjorie Sarnat   
17      Adult Coloring Book: Stress Relieving Patterns  Blue Star Coloring   
376  The Four Agreements: A Practical Guide to Pers...     Don Miguel Ruiz   
16   Adult Coloring Book: Stress Relieving Animal D...  Blue Star Coloring   
253              Old School (Diary of a Wimpy Kid #10)         Jeff Kinney   

     user_rating  reviews  price  year        genre  
507          4.8    26234      0  2015      Fiction  
210          4.6     6990      4  2015  Non Fiction  
206          4.5     3673      4  2015  Non Fiction  
111          4.7    17323      4  2015  Non Fiction  
123          4.8    14038      4  2015      Fiction  
54           4.8     4022      4  2015  Non Fiction  
63           4.8    10922      5  2015      Fiction  
89           4.6     2134      5  2015  Non Fiction  
478          4.9    19546      5  2015      Fiction  
201          4.6     5235      5  2015  Non Fiction  
28           4.6     5360      5  2015  Non Fiction  
224          4.8    18613      5  2015      Fiction  
55           4.8     3871      5  2015  Non Fiction  
17           4.4     2951      6  2015  Non Fiction  
376          4.7    23308      6  2015  Non Fiction  
16           4.6     2925      6  2015  Non Fiction  
253          4.8     6169      7  2015      Fiction

Агрегирование данных и соединения таблиц#

Последняя секция этого домашнего задания включает чуть более продвинутые функции. Но не волнуйтесь, pandas делает все операции простыми и понятными.

Для начала давайте посмотрим на максимальную и минимальную цены для каждого из жанров (используйте функции groupby и agg, для подсчета минимальных и максимальных значений используйте max и min). Не берите все столбцы, виберете только нужные вам

1..Ответ: Максимальная цена для жанра Fiction: **Ответ: 82**

2..Ответ: Минимальная цена для жанра Fiction: **Ответ: 0**

3..Ответ: Максимальная цена для жанра Non Fiction: **Ответ: 105**

4..Ответ: Минимальная цена для жанра Non Fiction: **Ответ: 0**

Теперь создайте новый датафрейм, который будет вмещать в себя количество книг для каждого из авторов (используйте функции groupby и agg, для подсчета количества используйте count). Не берите все столбцы, виберете только нужные

5.. Ответ: Какой размерности вышла таблица? **Ответ: (248, 1)**

6..Ответ: У какого автора больше всего книг? **Ответ: Jeff Kinney**

7..Ответ: Сколько книг у этого автора? **Ответ: 12**

In [317]:
x=df.groupby("genre").agg({"price":["min", "max"]})
#x = x.agg({"price":["min", "max"]})
print(f"1.2.3.4{x}")


1.2.3.4            price     
              min  max
genre                 
Fiction         0   82
Non Fiction     0  105


In [318]:
ndf=df.groupby("author").agg({"author":["count"]})
print(f"5.....{ndf.shape}")
ndf=pd.DataFrame(ndf)
a=ndf.max()
a=int(a[0])
for i, j in ndf.items():
  for k in range(0, ndf.shape[0]):    
    if j[k] == a:
      print(f"6.....{ndf.iloc[k],j[k]}")

5.....(248, 1)
6.....(author  count    12
Name: Jeff Kinney, dtype: int64, 12)


In [319]:
print(f'7.....{df[df["author"]=="Jeff Kinney"].agg({"author":["count"]})}')




7.....       author
count      12


Теперь создайте второй датафрейм, который будет вмещать средний рейтинг для каждого автора (используйте функции groupby и agg, для подсчета среднего значения используйте mean). Не берите все столбцы, виберете только нужные

8.. Ответ: У какого автора средний рейтинг минимальный? **Ответ: Donna Tartt**

9..Ответ: Какой у этого автора средний рейтинг? **Ответ: 3.9**

In [320]:
y=df.groupby("author").agg({"user_rating":["mean"]})
ndf_new=pd.DataFrame(y)

for i, j in y.items():
  for k in range(0, ndf_new.shape[0]):    
    if j[k] == y.min()[0]:
      print(f"8.....{ndf_new.iloc[k],j[k]}")

8.....(user_rating  mean    3.9
Name: Donna Tartt, dtype: float64, 3.9)


Соедините последние два датафрейма так, чтобы для каждого автора было видно количество книг и средний рейтинг (Используйте функцию concat с параметром axis=1). Сохраните результат в переменную Отсортируйте датафрейм по возрастающему количеству книг и возрастающему рейтингу (используйте функцию sort_values)

10..Ответ: Какой автор первый в списке? **Ответ: Muriel Barbery**

In [370]:
ndf_=df.groupby("author").agg({"author":["count"]})
ndf_.columns = ndf_.columns.droplevel()
concat_df=pd.concat([ndf_, ndf_new], axis=1)
concat_df.sort_values(["count", "mean"])

count      mean
author                                             
Muriel Barbery                          1  4.000000
Chris Cleave                            1  4.100000
Ian K. Smith M.D.                       1  4.100000
Pierre Dukan                            1  4.100000
Elizabeth Strout                        1  4.200000
...                                   ...       ...
American Psychological Association     10  4.500000
Suzanne Collins                        11  4.663636
Gary Chapman                           11  4.736364
Rick Riordan                           11  4.772727
Jeff Kinney                            12  4.800000

[248 rows x 2 columns]